### PR TITLE
PerfromActionOnMachineGroup Task changes

### PR DESCRIPTION
**Pull request by pankagar13**

### CODE REVIEW

File: Tasks/PerformActionOnMachineGroup/AzureHelper.ps1 [hunk 1]
```diff
@@ -71,26 +71,41 @@ function Initialize-AzureHelper
...
-            Write-Verbose "SubscriptionName : $subscriptionName" -Verbose
-            Write-Verbose "Username : $username" -Verbose
-            $securePassword = ConvertTo-SecureString $password -AsPlainText -Force
-            $psCredential = New-Object System.Management.Automation.PSCredential ($username, $securePassword)
-            $azureAccount = Add-AzureAccount -Credential $psCredential
-            if(!$azureAccount)
+            $serviceEndpoint = Get-ServiceEndpoint -Name $subscriptionId -Context $distributedTaskContext
+            if ($serviceEndpoint -eq $null)
+            {
+                throw (Get-LocalizedString -Key "A Connected Service with Id '{0}' could not be found. Ensure that this Connected Service was successfully provisioned using services tab in Admin UI" -ArgumentList $subscriptionId)
```
Comment (pankajkhanzode) [inline Tasks/PerformActionOnMachineGroup/AzureHelper.ps1]: Where is the resource for this string? - A Connected Service with Id '{0}' could not be found. Ensure that this

**Comment by pankajkhanzode** [comment]
Signing off.

File: Tasks/PerformActionOnMachineGroup/AzureHelper.ps1 (hunk 1 above)
Comment (pankagar13) [inline Tasks/PerformActionOnMachineGroup/AzureHelper.ps1]: I have provided -ArgumentList $subscriptionId in the end

File: Tasks/PerformActionOnMachineGroup/AzureHelper.ps1 [hunk 2]
```diff
@@ -71,26 +71,41 @@ function Initialize-AzureHelper
         $providerDataName = $machineGroup.ProviderDataList[0].Name
         Write-Verbose "Getting providerData : $providerDataName" -Verbose
         $providerData = Get-ProviderData -ProviderDataName $providerDataName -Connection $connection
-        $subscriptionName = $providerData.Properties.GetProperty("SubscriptionName")     
-        $username = $providerData.Properties.GetProperty("Username")
-        $password = $providerData.Properties.GetProperty("Password")
-
-        if( ![string]::IsNullOrEmpty($subscriptionName) -and ![string]::IsNullOrEmpty($username) -and ![string]::IsNullOrEmpty($password) )
+        $subscriptionId = $providerData.Properties.GetProperty("SubscriptionId")
```
Comment (abhinitmodi) [inline Tasks/PerformActionOnMachineGroup/AzureHelper.ps1]: There is a dll called Dtl.common where all the constants are defined. Please move "SubscriptionId" there

File: Tasks/PerformActionOnMachineGroup/AzureHelper.ps1 [hunk 3]
```diff
@@ -71,26 +71,41 @@ function Initialize-AzureHelper
...
+        if( ![string]::IsNullOrEmpty($subscriptionId) )
         {
-            Write-Verbose "SubscriptionName : $subscriptionName" -Verbose
-            Write-Verbose "Username : $username" -Verbose
-            $securePassword = ConvertTo-SecureString $password -AsPlainText -Force
-            $psCredential = New-Object System.Management.Automation.PSCredential ($username, $securePassword)
-            $azureAccount = Add-AzureAccount -Credential $psCredential
-            if(!$azureAccount)
+            $serviceEndpoint = Get-ServiceEndpoint -Name $subscriptionId -Context $distributedTaskContext
+            if ($serviceEndpoint -eq $null)
```
Comment (abhinitmodi) [inline Tasks/PerformActionOnMachineGroup/AzureHelper.ps1]: if($serviceEndPoint)

File: Tasks/PerformActionOnMachineGroup/AzureHelper.ps1 [hunk 4]
```diff
@@ -71,26 +71,41 @@ function Initialize-AzureHelper
...
+            {
+                throw (Get-LocalizedString -Key "A Connected Service with Id '{0}' could not be found. Ensure that this Connected Service was successfully provisioned using services tab in Admin UI" -ArgumentList $subscriptionId)
+            }
+            if ($serviceEndpoint.Authorization.Scheme -eq 'UserNamePassword')
```
Comment (abhinitmodi) [inline Tasks/PerformActionOnMachineGroup/AzureHelper.ps1]: Use constant for UserNamePassword and move it to constants dll

File: Tasks/PerformActionOnMachineGroup/AzureHelper.ps1 [hunk 5]
```diff
@@ -71,26 +71,41 @@ function Initialize-AzureHelper
...
+
+                Write-Verbose "SubscriptionName : $subscriptionName" -Verbose
+                Write-Verbose "Username : $username" -Verbose
+
+                $securePassword = ConvertTo-SecureString $password -AsPlainText -Force
+                $psCredential = New-Object System.Management.Automation.PSCredential ($username, $securePassword)
+                $azureAccount = Add-AzureAccount -Credential $psCredential
+                if(!$azureAccount)
+                {
+                    throw "There was an error with the Azure credentials used for machine group deployment"
```
Comment (abhinitmodi) [inline Tasks/PerformActionOnMachineGroup/AzureHelper.ps1]: Localize

File: Tasks/PerformActionOnMachineGroup/task.json [hunk 6]
```diff
@@ -22,7 +22,7 @@
             "type": "string", 
             "label": "Machine Group",
             "required": true,
-            "helpMarkDown": "Name of the Machine Group on which the operation is to be performed."
+            "helpMarkDown": "Name of the target Machine Group for performing the action."
```
Comment (abhinitmodi) [inline Tasks/PerformActionOnMachineGroup/task.json]: maintain consistent casing of Machine Group in the help mark down. Follow sentence case

File: Tasks/PerformActionOnMachineGroup/AzureHelper.ps1 [hunk 7]
```diff
@@ -71,26 +71,41 @@ function Initialize-AzureHelper
...
+                throw (Get-LocalizedString -Key "A Connected Service with Id '{0}' could not be found. Ensure that this Connected Service was successfully provisioned using services tab in Admin UI" -ArgumentList $subscriptionId)
+            }
+            if ($serviceEndpoint.Authorization.Scheme -eq 'UserNamePassword')
             {
-                throw "There was an error with the Azure credentials used for machine group deployment"
+                $username = $serviceEndpoint.Authorization.Parameters.UserName
+                $password = $serviceEndpoint.Authorization.Parameters.Password
+                $subscriptionName = $serviceEndpoint.Data.SubscriptionName
+
+                Write-Verbose "SubscriptionName : $subscriptionName" -Verbose
```
Comment (abhinitmodi) [inline Tasks/PerformActionOnMachineGroup/AzureHelper.ps1]: Establishing connection to subscription $subscriptionName using the identity $username.

Remove both the write-verbose and add this here. Add Write-Host, Make sure to localize it

File: Tasks/PerformActionOnMachineGroup/AzureHelper.ps1 (hunk 5 above)
Comment (abhinitmodi) [inline Tasks/PerformActionOnMachineGroup/AzureHelper.ps1]: The credentials with which the connected service $subscriptionName is configured are invalid.

File: Tasks/DeployAzureResourceGroup/DtlServiceHelper.ps1 [hunk 8]
```diff
@@ -20,8 +20,10 @@ function Create-ProviderData
     
     Write-Verbose "Registering provider data $providerDataName" -Verbose
 
-    $propertyBag = Get-ServiceEndPointDetails -ConnectedServiceName $ConnectedServiceName
-
+    $propertyBag = New-Object 'System.Collections.Generic.Dictionary[string, Microsoft.VisualStudio.Services.DevTestLabs.Model.PropertyBagData]'        
+    $property = New-Object Microsoft.VisualStudio.Services.DevTestLabs.Model.PropertyBagData($false, $subscriptionId)
+    $propertyBag.Add("SubscriptionId", $property)
```
Comment (abhinitmodi) [inline Tasks/DeployAzureResourceGroup/DtlServiceHelper.ps1]: Constants fiel in common dll

Comment (pankagar13) [inline Tasks/DeployAzureResourceGroup/DtlServiceHelper.ps1]: as we discussed with anubhav we will add those thigns later in .psm1 module

**Comment by bryanmacfarlane** [comment]
pankagar13 - please accept or close

**Comment by kmkumaran** [comment]
Old PR. Closing it.

**Pull request closed without merging**